torchtitan/models/deepseek_v3/model/state_dict_adapter.py

```diff
@@ -52,6 +60,278 @@ def __init__(self, model_args: DeepSeekV3ModelArgs, hf_assets_path: str | None):
+
+        # Step 3: Create new placements for individual expert weights
+        new_placements = []
+        for i, name in enumerate(device_mesh.mesh_dim_names):
+            placement = dtensor_placements[i]
+            if placement.dim == 0:
+                # Convert dim-0 sharding to replication for individual experts
+                new_placements.append(Replicate())
+            elif isinstance(placement, Shard):
+                # Keep other shard dimensions (individual expert weight has 2D)
+                new_placements.append(Shard(placement.dim))
+            elif isinstance(placement, _StridedShard):
+                # Keep strided shard with same parameters
+                new_placements.append(
+                    _StridedShard(placement.dim, placement.split_factor)
+                )
+            else:
+                raise ValueError(f"Unsupported placement type: {type(placement)}")
+
+        # Step 4: Create individual expert DTensors
+        assert isinstance(
+            grouped_expert_weight, DTensor
+        ), "Expected DTensor for grouped expert weight"
+
+        local_grouped_weights = grouped_expert_weight._local_tensor
+        expected_local_experts = end_index - start_index
+
+        if local_grouped_weights.shape[0] != expected_local_experts:
+            raise ValueError(
+                f"Local tensor shape mismatch: expected {expected_local_experts} experts, "
+                f"got {local_grouped_weights.shape[0]}"
+            )
+
+        local_expert_tensors = {}
+        for expert_id in range(start_index, end_index):
+            expert_key = abstract_key.format(layer_id, expert_id)
+            local_expert_index = expert_id - start_index
+
+            # Extract individual expert weight and add batch dimension temporarily
+            expert_weight = local_grouped_weights[local_expert_index, :, :].unsqueeze(0)
+
+            # Create DTensor and remove batch dimension (experts dimension is removed)
+            expert_dtensor = DTensor.from_local(
+                expert_weight, device_mesh, new_placements, run_check=False
+            ).squeeze(0)
+
+            local_expert_tensors[expert_key] = expert_dtensor
+
+        return local_expert_tensors
+
+    def _concatenate_local_expert_weights(
+        self,
+        expert_weights_by_layer: dict[str, Any],
+        abstract_key: str,
+        device_mesh: DeviceMesh,
+    ) -> torch.Tensor:
+        """
+        Try to concatenate the weights of separate experts into GroupedExperts weights.
+        """
+        for layer in expert_weights_by_layer.keys():
+            # If we have all the experts for this abstract_key, concatenate them
+            experts = expert_weights_by_layer[layer][abstract_key]
+            expected_n_experts = (
+                self.local_experts_indices[abstract_key][1]
+                - self.local_experts_indices[abstract_key][0]
+            )
+            if len(experts) == expected_n_experts:
```

Comment:
I want to return instead of continue, but I'm not sure if that is true. Also see the comment below. I want to understand if we can avoid looping layers.
```suggestion
            if len(experts) < expected_n_experts:
            	continue
```

Reply:
When a new fqn is processed, the `_concatenate_local_expert_weights ` function is called to check if we can concatenate the individual experts weights into GroupedExpert weights. If the fqn has `layer_num`, the only possible layer can be merged is layer with id=`layer_num`. In this way, we could remove the loop of layers as you suggested. 

So we could use `return` here instead of `continue` after remove the loop